crates/net/network/src/transactions.rs

```diff
@@ -1074,116 +1271,341 @@ struct Peer {
+            );
+
+            let Some((retries, peers)) = self.unknown_hashes.get(&hash) else { return };
+
+            if let Some(peer_id) = fallback_peer {
+                // peer has not yet requested hash
+                peers.insert(peer_id);
+            } else {
+                // peer in caller's context has requested hash and is hence not eligible as
+                // fallback peer.
+                if *retries >= MAX_REQUEST_RETRIES_PER_TX_HASH {
+                    warn!(target: "net::tx",
+                        hash=format!("{hash:#}"),
+                        retries=retries,
+                        "retry limit for `GetPooledTransactions` requests reached for hash, dropping hash"
+                    );
+                    max_retried_hashes.push(hash);
+                    continue;
+                }
+                *retries += 1;
+            }
+            if let (_, Some(evicted_hash)) = self.buffered_hashes.insert_and_get_evicted(hash) {
+                _ = self.unknown_hashes.remove(&evicted_hash);
+                _ = self.eth68_meta.remove(&evicted_hash);
+            }
+        }
+
+        self.remove_from_unknown_hashes(max_retried_hashes);
     }
 
     /// Removes the provided transaction hashes from the inflight requests set.
     ///
     /// This is called when we receive full transactions that are currently scheduled for fetching.
     #[inline]
-    fn on_received_full_transactions_broadcast<'a>(
+    fn on_received_full_transactions_broadcast(
+        &mut self,
+        hashes: impl IntoIterator<Item = TxHash>,
+    ) {
+        self.remove_from_unknown_hashes(hashes)
+    }
+
+    fn filter_unseen_hashes(
         &mut self,
-        hashes: impl IntoIterator<Item = &'a TxHash>,
+        new_announced_hashes: &mut Vec<TxHash>,
+        peer_id: PeerId,
+        is_session_active: impl Fn(PeerId) -> bool,
     ) {
-        self.remove_inflight_hashes(hashes)
+        // filter out inflight hashes, and register the peer as fallback for all inflight hashes
+        new_announced_hashes.retain(|hash| {
+            // occupied entry
+            if let Some((_retries, ref mut backups)) = self.unknown_hashes.peek_mut(hash) {
+                // hash has been seen but is not inflight
+                if self.buffered_hashes.remove(hash) {
+                    return true
+                }
+                // hash has been seen and is in flight. store peer as fallback peer.
+                //
+                // remove any ended sessions
+                let mut ended_sessions = vec!();
+                for &peer_id in backups.iter() {
+                    if is_session_active(peer_id) {
+                        ended_sessions.push(peer_id);
+                    }
+                }
+                for peer_id in ended_sessions {
+                    backups.remove(&peer_id);
+                }
+                backups.insert(peer_id);
+                return false
```

Comment:
I see, we need cleanup old peers somewhere, I believe this is fine because alternate peers are capped

Reply:
exactly, this happens on-op:

1) when we store a previously unseen hash to unknown hashes, then we make sure we don't evict an alive peer from the cache in favour of a lru but dead peer
2) when we try to find an idle peer for a hash. incase we have to loop through those fallback peers again when retrying for that hash, we spared ourselves an iteration per removed fallback peer. rn the default number of fallback peers is used which is set to 2, same with retries which is also low, set to 2, but a node operator could compile with a bigger capacity for fallback peers and more retries, and then it pays off.

prior to these situations, (1) and (2), it doesn't bother us if there are dead fallback peers since the list is bounded, as you point out. the operation of looping through all nested lists of fallback peers in unknown-hashes, to check if a peer is in it, is kind of costly. these to reasons makes it best to skip removing the peer from fallbacks when we get `NetworkEvent::SessionClosed`.